internal/e2e: test flake TestRouteWithAServiceWeight

--- FAIL: TestRouteWithAServiceWeight (0.00s)
    rds_test.go:1710:

fails because the order of services with the same name and different weights is unstable.

--- a/internal/contour/route.go
+++ b/internal/contour/route.go
@@ -249,45 +249,27 @@ func prefixmatch(prefix string) route.RouteMatch {
 // action computes the cluster route action, a *route.Route_route for the
 // supplied ingress and backend.
 func actionroute(services []*dag.Service, ws bool, timeout time.Duration) *route.Route_Route {
-	var totalWeight int
-	upstreams := []*route.WeightedCluster_ClusterWeight{}
-
-	// Loop over all the upstreams and add to slice
-	for _, svc := range services {
-		// Create the upstream
-		upstreams = append(upstreams, &route.WeightedCluster_ClusterWeight{
-			Name:   hashname(60, svc.Namespace(), svc.Name(), strconv.Itoa(int(svc.Port))),
-			Weight: &types.UInt32Value{Value: uint32(svc.Weight)},
-		})
-		totalWeight += svc.Weight
-	}
-
-	// sort upstreams by name to keep the list stable
-	sort.Stable(clusterWeightByName(upstreams))
-
-	// Check if no weights were defined, if not default to even distribution
-	if totalWeight == 0 {
-		for _, u := range upstreams {
-			u.Weight.Value = 1
-		}
-		totalWeight = len(services)
-	}
-
 	rr := route.Route_Route{
 		Route: &route.RouteAction{
 			ClusterSpecifier: &route.RouteAction_WeightedClusters{
-				WeightedClusters: &route.WeightedCluster{
-					Clusters: upstreams,
-					TotalWeight: &types.UInt32Value{
-						Value: uint32(totalWeight),
-					},
-				},
+				WeightedClusters: weightedclusters(services),
 			},
 		},
 	}
+
+	// Check if no weights were defined, if not default to even distribution
+	clusters := rr.Route.ClusterSpecifier.(*route.RouteAction_WeightedClusters).WeightedClusters
+	if clusters.TotalWeight.Value == 0 {
+		for _, c := range clusters.Clusters {
+			c.Weight.Value = 1
+		}
+		clusters.TotalWeight.Value = uint32(len(clusters.Clusters))
+	}
+
 	if ws {
 		rr.Route.UseWebsocket = &types.BoolValue{Value: ws}
 	}
+
 	switch timeout {
 	case 0:
 		// no timeout specified, do nothing
@@ -303,11 +285,34 @@ func actionroute(services []*dag.Service, ws bool, timeout time.Duration) *route
 	return &rr
 }
 
+func weightedclusters(services []*dag.Service) *route.WeightedCluster {
+	var wc route.WeightedCluster
+	var total int
+	for _, svc := range services {
+		total += svc.Weight
+		wc.Clusters = append(wc.Clusters, &route.WeightedCluster_ClusterWeight{
+			Name:   hashname(60, svc.Namespace(), svc.Name(), strconv.Itoa(int(svc.Port))),
+			Weight: &types.UInt32Value{Value: uint32(svc.Weight)},
+		})
+	}
+	wc.TotalWeight = &types.UInt32Value{
+		Value: uint32(total),
+	}
+	sort.Stable(clusterWeightByName(wc.Clusters))
+	return &wc
+}
+
 type clusterWeightByName []*route.WeightedCluster_ClusterWeight
 
-func (c clusterWeightByName) Len() int           { return len(c) }
-func (c clusterWeightByName) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }
-func (c clusterWeightByName) Less(i, j int) bool { return c[i].Name < c[j].Name }
+func (c clusterWeightByName) Len() int      { return len(c) }
+func (c clusterWeightByName) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
+func (c clusterWeightByName) Less(i, j int) bool {
+	if c[i].Name == c[j].Name {
+		return c[i].Weight.Value < c[j].Weight.Value
+	}
+	return c[i].Name < c[j].Name
+
+}
 
 // hashname takes a lenth l and a varargs of strings s and returns a string whose length
 // which does not exceed l. Internally s is joined with strings.Join(s, "/"). If the

--- a/internal/e2e/rds_test.go
+++ b/internal/e2e/rds_test.go
@@ -1699,8 +1699,9 @@ func TestRouteWithAServiceWeight(t *testing.T) {
 					Port:   80,
 					Weight: 90,
 				}, {
-					Name: "kuard",
-					Port: 80, Weight: 60,
+					Name:   "kuard",
+					Port:   80,
+					Weight: 60,
 				}},
 			}},
 		},
@@ -1713,8 +1714,8 @@ func TestRouteWithAServiceWeight(t *testing.T) {
 		Routes: []route.Route{{
 			Match: prefixmatch("/a"), // match all
 			Action: routeweightedcluster(
-				weightedcluster{"default/kuard/80", 90},
 				weightedcluster{"default/kuard/80", 60},
+				weightedcluster{"default/kuard/80", 90},
 			),
 		}},
 	}}, nil)

